### PR TITLE
Concurrent cascade delete operations leading to confusing errors

### PR DESCRIPTION
**Pull request by lukedegruchy**
Closes [3889](https://github.com/hapifhir/hapi-fhir/issues/3889)

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	public static final int RETRY_MAX_ATTEMPTS = 4;
+
+	private static final Logger ourLog = LoggerFactory.getLogger(SafeDeleter.class);
+	private final DaoRegistry myDaoRegistry;
+	private final IInterceptorBroadcaster myInterceptorBroadcaster;
+	private final TransactionTemplate myTxTemplate;
+
+	private final RetryTemplate myRetryTemplate = getRetryTemplate();
+
+	public SafeDeleter(DaoRegistry theDaoRegistry, IInterceptorBroadcaster theInterceptorBroadcaster, PlatformTransactionManager thePlatformTransactionManager) {
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java:42]: should this be a bean?  If so, I'd rename it to `SafeDeleterSvc`

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+			}
+		}
+
+		return retVal;
+	}
+
+	private Integer handleNextSource(RequestDetails theRequest, DeleteConflictList theConflictList, TransactionDetails theTransactionDetails, DeleteConflict next, IdDt nextSource, String nextSourceId) {
+		IFhirResourceDao<?> dao = myDaoRegistry.getResourceDao(nextSource.getResourceType());
+
+		return myRetryTemplate.execute(retryContext -> {
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java:69]: add a comment to mention that ResourceVersionConflictException will be retried

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+			if (!cascadeDeleteIdCache.contains(nextSourceId)) {
+				cascadeDeleteIdCache.add(nextSourceId);
+				retVal += handleNextSource(theRequest, theConflictList, theTransactionDetails, next, nextSource, nextSourceId);
+			}
+		}
+
+		return retVal;
+	}
+
+	private Integer handleNextSource(RequestDetails theRequest, DeleteConflictList theConflictList, TransactionDetails theTransactionDetails, DeleteConflict next, IdDt nextSource, String nextSourceId) {
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java:66]: javadoc returns number of resources deleted

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/delete/SafeDeleterTest.java [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+		myCascadeDeleteInterceptor.awaitExpected();
+
+		//   Let's delete the second patient from under its nose.
+		ourLog.info("delete patient 2");
+		myPatientDao.delete(patient2Id);
+
+		// Unpause and delete the first patient
+		myCascadeDeleteInterceptor.release("first");
+
+		assertEquals(1, future.get());
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/delete/SafeDeleterTest.java:128]: comment this returns the total number of resources deleted

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	private final RetryTemplate myRetryTemplate = getRetryTemplate();
+
+	public SafeDeleter(DaoRegistry theDaoRegistry, IInterceptorBroadcaster theInterceptorBroadcaster, PlatformTransactionManager thePlatformTransactionManager) {
+		myDaoRegistry = theDaoRegistry;
+		myInterceptorBroadcaster = theInterceptorBroadcaster;
+		myTxTemplate = new TransactionTemplate(thePlatformTransactionManager);
+		myTxTemplate.setPropagationBehavior(TransactionTemplate.PROPAGATION_REQUIRES_NEW);
+	}
+
+	public Integer delete(RequestDetails theRequest, DeleteConflictList theConflictList, TransactionDetails theTransactionDetails) {
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java:49]: Javadoc returns the total number of resources deleted

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/delete/SafeDeleterTest.java [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+		// Patient 1 We are paused after reading the version but before deleting the first patient.
+		myCascadeDeleteInterceptor.awaitExpected();
+
+
+		// Unpause and delete the first patient
+		myCascadeDeleteInterceptor.setExpectedCount(1);
+		myCascadeDeleteInterceptor.release("first");
+
+		// Patient 2 We are paused after reading the version but before deleting the first patient.
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/delete/SafeDeleterTest.java:161]:
```suggestion
		// Patient 2 We are paused after reading the version but before deleting the second patient.
```

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/config/TestR4WithDelayConfig.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+@Configuration
+@Import({
+	JpaR4Config.class,
+	HapiJpaConfig.class,
+	TestJPAConfig.class,
+	TestHSearchAddInConfig.DefaultLuceneHeap.class,
+	JpaBatch2Config.class,
+	Batch2JobsConfig.class
+})
+public class TestR4WithDelayConfig extends TestR4Config {
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/config/TestR4WithDelayConfig.java:53]: this appears to be unused.  recommend deleting

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/config/DelayListener.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
+package ca.uhn.fhir.jpa.test.config;
+
+import net.ttddyy.dsproxy.ExecutionInfo;
+import net.ttddyy.dsproxy.QueryInfo;
+import net.ttddyy.dsproxy.support.ProxyDataSourceBuilder;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class DelayListener implements ProxyDataSourceBuilder.SingleQueryExecution {
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/config/DelayListener.java:10]: this appears to be unused. recommend deleting

Comment (lukedegruchy) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/config/DelayListener.java:10]: This is used by the CascadingDeleteInterceptorMultiThreadTest

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/config/TestR4WithDelayConfig.java (hunk 7 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/config/TestR4WithDelayConfig.java:53]: This is used by the CascadingDeleteInterceptorMultiThreadTest

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java (hunk 1 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java]: Done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java (hunk 2 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java]: Done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java (hunk 3 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java]: Done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java (hunk 5 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleter.java]: Done

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/delete/SafeDeleterTest.java (hunk 6 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/delete/SafeDeleterTest.java]: Done

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/delete/SafeDeleterTest.java (hunk 4 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/delete/SafeDeleterTest.java]: Done.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/interceptor/CascadingDeleteInterceptor.java [hunk 9]
```diff
@@ -118,36 +116,13 @@ public DeleteConflictOutcome handleDeleteConflicts(DeleteConflictList theConflic
...
-			}
-		}
+		mySafeDeleterSvc.delete(theRequest, theConflictList, theTransactionDetails);
 
 		return new DeleteConflictOutcome().setShouldRetryCount(MAX_RETRY_ATTEMPTS);
 	}
 
-	@SuppressWarnings("unchecked")
-	private List<String> getCascadedDeletesMap(RequestDetails theRequest, boolean theCreate) {
+	// TODO LUKE move this
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/interceptor/CascadingDeleteInterceptor.java:124]: Nitpick: Is this intended to be moved in this PR? If not, recommend fleshing this TODO out a bit more so you can remember what the intention was (ie where should it move, and why?)

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleterSvc.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+import org.springframework.retry.backoff.FixedBackOffPolicy;
+import org.springframework.retry.policy.SimpleRetryPolicy;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.transaction.support.TransactionTemplate;
+
+import java.util.Collections;
+import java.util.List;
+
+public class SafeDeleterSvc {
```
[the remark this answers is not in the document]
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/delete/SafeDeleterSvc.java]: Done

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/config/DelayListener.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+	}
+
+	@Override
+	public void execute(ExecutionInfo execInfo, List<QueryInfo> queryInfoList) {
+		if (enabled && queryInfoList.get(0).getQuery().contains("from HFJ_RES_LINK")) {
+			if (deleteCount.getAndIncrement() == 0) {
+				try {
+					Thread.sleep(500L);
+				} catch (InterruptedException theE) {
+					theE.printStackTrace();
```
[the remark this answers is not in the document]
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/config/DelayListener.java]: Done.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/RestfulServer.java [hunk 12]
```diff
@@ -1190,6 +1190,7 @@ protected void handleRequest(RequestTypeEnum theRequestType, HttpServletRequest
 			writeExceptionToResponse(theResponse, e);
 
 		} catch (Throwable e) {
+//			ourLog.error("LUKE: Exception on RestfulServer!", e);
```
[the remark this answers is not in the document]
Comment (lukedegruchy) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/RestfulServer.java]: Done.

**Review by jamesagnew — APPROVED** [review]

**Pull request merged**